src/Symfony/Component/HttpClient/NativeHttpClient.php

```diff
@@ -0,0 +1,400 @@
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\HttpClient\Exception\InvalidArgumentException;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpClient\Response\NativeResponse;
+use Symfony\Component\HttpClient\Response\ResponseStream;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\HttpClientTrait;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * A portable implementation of the HttpClientInterface contracts based on PHP stream wrappers.
+ *
+ * PHP stream wrappers are able to fetch response bodies concurrently,
+ * but each request is opened synchronously.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class NativeHttpClient implements HttpClientInterface
+{
+    use HttpClientTrait;
+
+    private $defaultOptions = self::OPTIONS_DEFAULTS;
+    private $noProxy;
+    private $multi;
+
+    /**
+     * @param array $defaultOptions     Default requests' options
+     * @param int   $maxHostConnections The maximum number of connections to open
+     * @param int   $noProxy            A comma separated list of hosts that do not require a proxy to be reached, even if one is specified
+     */
+    public function __construct(array $defaultOptions = [], int $maxHostConnections = 6, string $noProxy = null)
+    {
+        if ($defaultOptions) {
+            [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, self::OPTIONS_DEFAULTS);
+        }
+
+        $noProxy = $noProxy ?? $_SERVER['no_proxy'] ?? $_SERVER['NO_PROXY'] ?? '';
+        $this->noProxy = $noProxy ? preg_split('/[\s,]+/', $noProxy) : [];
+
+        // Use an internal stdClass object to share state between the client and its responses
+        $this->multi = (object) [
+            'openHandles' => [],
+            'handlesActivity' => [],
+            'pendingResponses' => [],
+            'maxHostConnections' => $maxHostConnections,
+            'responseCount' => 0,
+            'dnsCache' => [],
+            'handles' => [],
+            'sleep' => false,
+            'id' => random_int(PHP_INT_MIN, PHP_INT_MAX),
+        ];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, iterable $options = []): ResponseInterface
+    {
+        [$url, $options] = self::prepareRequest($method, $url, $options, $this->defaultOptions);
+
+        if ($options['bindto'] && file_exists($options['bindto'])) {
+            throw new TransportException(__CLASS__.' cannot bind to local Unix sockets, use e.g. CurlHttpClient instead.');
+        }
+
+        // Cast the body to string
+        if (\is_resource($body = $options['body'])) {
+            $options['body'] = stream_get_contents($body);
+        } elseif ($body instanceof \Closure) {
+            $options['body'] = '';
+
+            while ('' !== $data = $body(16372)) {
```

Comment:
Why do we pass `16372` as an argument? 🤔

Reply:
Because that's the signature we expect from these callable: they should accept a chunk size. curl uses that value, I decide to use the same.

Reply:
Maybe it would be better to define it as a constant so we don't rely on a magic number.

Reply:
done